packages/rum-core/src/common/truncate.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2017-present, Elasticsearch BV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+/**
+ * Goes over the object in Breadth first fashion
+ * Example:
+ * {
+ *    a: '10',
+ *    b: { c: 20, d: { e: 30 } },
+ *    f: { g: '100' }
+ * }
+ *
+ * Execute callbacks on below paths in order
+ * ['a'],  ['b', 'c'], ['b', 'd', 'e'], ['f', 'g']
+ */
+
+function breathFilter(obj = {}, callback, path = []) {
+  const queue = []
+  let index = 0
+  const keys = Object.keys(obj)
+  if (keys.length === index) {
+    return obj
+  }
+  queue.push(keys[index])
+
+  while (queue.length > 0) {
+    const currKey = queue.shift()
+    const value = obj[currKey]
+
+    if (typeof value === 'object') {
+      breathFilter(value, callback, path.concat(currKey))
+    } else {
+      obj[currKey] = callback(value, path.concat(currKey))
+    }
+    index++
+    if (index < keys.length) {
+      queue.push(keys[index])
+    }
+  }
+  return obj
+}
+
+function truncate(value, limit, required = false, placeholder = 'N/A') {
+  if (required && !value) {
+    value = placeholder
+  }
+  if (value) {
+    return String(value).substr(0, limit)
+  }
+  return value
+}
+
+function truncateMetadata(metadata, opts) {
+  return breathFilter(metadata, (value, path) => {
+    if (typeof value !== 'string') {
+      return value
+    }
+    let limit
+    let required = false
+    switch (path[0]) {
```

Comment:
maybe we should change these switch case to data models that we define for each field something like

```
var fieldSpec = {
'service.version': { required: false, limit:1024 }
}
```

We can also reverse the logic so instead of going over all fields of the payload we go over the `fieldSpec` fields and apply the truncation necessary. This could have a performance advantage since we don't need to visit all fields of the payload. This would also resolve the requirement of different limit depending on the keyword/non-keyword status.

I'm happy to make a prototype to show how it could work.

Reply:
In either case the algorithm would be O(N), I am pretty sure we wont again much in performance . by removing switch statements. I am happy to be proven wrong here. 

Plus, the solution you are proposing has a disadvantage of maintaining arbitrary objects for each field vs variables.

Reply:
I tried prototyping the idea you mentioned here, So in the end what did change was the removal of breadthFirst function that I wrote which exactly spits out the arrays with comma separated format like `['a'],  ['b', 'c'], ['b', 'd', 'e'], ['f', 'g']` on an object of the form 

```js
  {
     a: '10',
     b: { c: 20, d: { e: 30 } },
     f: { g: '100' }
  }

// `['a'],  ['b', 'c'], ['b', 'd', 'e'], ['f', 'g']`
```

If we have our data models for transformations like this 
```js
{
   a:  { limit: stringLimit, required: false }
   'b.c': { limit: stringLimit, required: false },
   'b.d.e': { limit: stringLimit, required: true },
}
```
vs

```js
switch (path[0]) {
   case: "b": 
      switch (path[1]) {
          case: "c": 
          case: "d": 
             required = true
             limit = stringLimit
          break
      }
}
```

+ The switch case version looks more cleaner logically as you can visualize the depth in a cleaner way and provides logical grouping as opposed to the dot separated ones. 
+ I am not in favor of the second solution since it creates arbitrary objects which comes at its own perf cost. Since we run on the browser, I would prefer creating minimal intermediate objects since it can accumulate in to GC pressure. 
+ I did a perf profile of the switch version with traversing the dot separated object version, Literally not even +/- 2ms difference since it both requires traversal of the objects atleast once. The advantage on dot level comes only if the data model is small which comes only if we want to skip certain spans/transaction with lots of context information which is not applicable in an ideal scenario.

Reply:
The arbitrary object is only created once and is never changed, I don't think it's cost is very significant, why do you think otherwise?

Furthermore, my main point is that I can create a hugely nested payload and the current solution would traverse all of it, even if the object is polluted with fields that are not even part of the final payload. This actually happens when libraries or applications change object prototype, I know it's a bad practice but it happens. Why should we allow the shape of the object that we are traversing affect our algorithm while we actually have a constant and pre-defined number of fields that we can traverse?